ospfd/ospf_vty.c

```diff
@@ -6111,86 +6384,176 @@ static void show_lsa_detail_proc(struct vty *vty, struct route_table *rt,
 		break;
 	}
+
+	if (json)
+		json_object_object_add(json, show_database_desc[type],
+				       json_lsa_type);
 }
 
 static void show_lsa_detail_adv_router_proc(struct vty *vty,
 					    struct route_table *rt,
-					    struct in_addr *adv_router)
+					    struct in_addr *adv_router,
+					    json_object *json)
 {
 	struct route_node *rn;
 	struct ospf_lsa *lsa;
 
 	for (rn = route_top(rt); rn; rn = route_next(rn))
-		if ((lsa = rn->info))
+		if ((lsa = rn->info)) {
+			json_object *json_lsa = NULL;
+
 			if (IPV4_ADDR_SAME(adv_router,
 					   &lsa->data->adv_router)) {
 				if (CHECK_FLAG(lsa->flags, OSPF_LSA_LOCAL_XLT))
 					continue;
+				if (json)
+					json_lsa = json_object_new_object();
+
 				if (show_function[lsa->data->type] != NULL)
-					show_function[lsa->data->type](vty,
-								       lsa);
+					show_function[lsa->data->type](
+						vty, lsa, json_lsa);
+				if (json)
+					json_object_object_add(
+						json, inet_ntoa(lsa->data->id),
+						json_lsa);
 			}
+		}
 }
 
 /* Show detail LSA information. */
 static void show_lsa_detail_adv_router(struct vty *vty, struct ospf *ospf,
-				       int type, struct in_addr *adv_router)
+				       int type, struct in_addr *adv_router,
+				       json_object *json)
 {
 	struct listnode *node;
 	struct ospf_area *area;
+	json_object *json_lstype = NULL;
+	json_object *json_area = NULL;
+
+	if (json)
+		json_lstype = json_object_new_object();
 
 	switch (type) {
 	case OSPF_AS_EXTERNAL_LSA:
 	case OSPF_OPAQUE_AS_LSA:
-		vty_out(vty, "                %s \n\n",
-			show_database_desc[type]);
+		if (!json)
+			vty_out(vty, "                %s \n\n",
+				show_database_desc[type]);
+
 		show_lsa_detail_adv_router_proc(vty, AS_LSDB(ospf, type),
-						adv_router);
+						adv_router, json_lstype);
 		break;
 	default:
+
 		for (ALL_LIST_ELEMENTS_RO(ospf->areas, node, area)) {
-			vty_out(vty, "\n                %s (Area %s)\n\n",
-				show_database_desc[type],
-				ospf_area_desc_string(area));
-			show_lsa_detail_adv_router_proc(
-				vty, AREA_LSDB(area, type), adv_router);
+			if (json)
+				json_area = json_object_new_object();
+			else
+				vty_out(vty,
+					"\n                %s (Area %s)\n\n",
+					show_database_desc[type],
+					ospf_area_desc_string(area));
+			show_lsa_detail_adv_router_proc(vty,
+							AREA_LSDB(area, type),
+							adv_router, json_area);
+
+			if (json)
+				json_object_object_add(
+					json_lstype,
+					ospf_area_desc_string(area), json_area);
 		}
 		break;
 	}
+
+	if (json)
+		json_object_object_add(json, show_database_desc[type],
+				       json_lstype);
+}
+
+/* Generate uniq LSA json key */
+static void ospf_genarate_json_lsa_key(char *key, size_t size,
+				       struct ospf_lsa *lsa)
+{
+	int len = 0;
+
+	/* LSA can be uniquly identifed/fecthed using LSID and
+	 * Advertiment router id.So LSID+ADVRTRID should be used
+	 * as key while storing in json object.
+	 */
+	len = snprintf(key, size, "%s-", inet_ntoa(lsa->data->id));
```

Comment:
Even though this looks fine, I think it would be best to switch to a safer idiom:

```c
snprintf(key, size, "%pI4-", &lsa->data->id);
strlcat(key, inet_ntoa(lsa->data->adv_router), size);
```

Reply:
snprintf  is not supported to this formatter . Used snprintfrr for this.